Add a section "Stub package versioning" (#10840)
python/typeshed: README.md

Type checkers should be able to use these stub packages when installed. For more
details, see the documentation for your type checker.

> ### Stub package versioning

Reviewer: ```suggestion
### Package versioning for third-party stubs
```

> Version numbers of third-party stub packages consist of at least four elements,
> where the last element represents the stub version. The other elements
> are the version of the package being stubbed.

Reviewer: ```suggestion
Version numbers of third-party stub packages consist of at least four parts.
All parts of the stub version, except for the last part, correspond to the
version of the runtime package being stubbed. For example, if the `types-foo`
package has version `1.1.2.0`, this signifies that the package contains stubs
for the `foo==1.1.2`. The final element of the version indicates that this is
the first revision of the stubs for `foo`: if an update to the stubs was
pushed the following day (but the stubs were still aiming to provide
annotations for `foo==1.1.2`), then the version of `types-foo` would
increment to `1.1.2.1`.
```

Author: I used a slightly modified version of your wording.

> While we try to minimize the number of breaking changes, due to the nature
> of stubs, every version can introduce changes that make your code fail
> to type check.

Reviewer: ```suggestion
At typeshed, we try to keep breaking changes to a minimum. However, due to the
nature of stubs, any version bump can introduce changes that might make your
code fail to type check.
```

> There are several strategies to pin the version of the stubs you are using,
> all with their own tradeoffs:

Reviewer: ```suggestion
There are several strategies available for specifying the version of a stubs
package you're using, each with its own tradeoffs
```

* Use the same pin that you use for the package being stubbed. For example,
  if you use `requests>=2.30.0,<2.32`, you can use
  `types-requests>=2.30.0,<2.32`. This ensures that the stubs are compatible
  with the package you are using, but it carries a small risk of breaking

>   type checking due to changes in the stubs. Also, stubs often lag behind

Reviewer: ```suggestion
  type checking due to changes in the stubs.
  
  Another risk of this strategy is that stubs often lag behind
```

  the package being stubbed. You might want to force the package being stubbed
  to a certain minimum version, because it fixes a critical bug, but you
  could be unable to update the stubs, since an update has not been released.
* When the package being stubbed is using [Semantic Versioning](https://semver.org/lang/de/),
  you could use the same major version for the stubs. For example, if you use
  `requests>=2.30.0,<2.33`, you can use `types-requests>=2.29.3.4,<3`. This
  means that the stubs are not as tighly coupled to the package being stubbed,
  and you will gain potential improvements in the stubs, but it carries a
  risk, since the stubs might provide a different API surface than the
  implementation. It also works around the problem that stubs often lag behind

>   the package being stubbed.

Reviewer: I'm not sure I fully understand this strategy :/

Author: I removed it, I didn't really understand it myself.

* Pin the stubs to a known good version and update manually. For example, if
  you use `types-requests==2.31.0.1`, you can be sure that upgrading
  dependencies will not break type checking. However, you will miss out on

>   improvements in the stubs that could potentially improve type checking.

Reviewer: Well, you'll only "miss out" for as long as you fail to upgrade. This depends on how often you bump your dependencies, right?

  It also has the risk that the stubs you are using are not compatible with
  the package being stubbed.
* Don't pin the stubs. This is the most flexible option, but it carries the
  risk that the stubs become incompatible with the package being stubbed.

>   For example, when a new major version of the package is released and the
>   stubs are updated before you update the package being stubbed.

Reviewer: ```suggestion
  For example, if a new major version of the package is released, there's a
  chance the stubs might be updated to reflect the new version of the runtime
  package before you update the package being stubbed.
```

You can also use the different strategies as needed. For example, you could
default match the pinning, but pin the stubs to a known good version when

> a problem arises that can't easily be fixed.

Reviewer: If we used a numbered list rather than a bulleted list for the strategies, we could say something like "you could default to strategy (1), but fall back to strategy (2) when a problem arises that can't easily be fixed". That might be slightly more elegant.

Type checkers should be able to use these stub packages when installed. For more
details, see the documentation for your type checker.

### Package versioning for third-party stubs

Version numbers of third-party stub packages consist of at least four parts.
All parts of the stub version, except for the last part, correspond to the
version of the runtime package being stubbed. For example, if the `types-foo`
package has version `1.1.2.0`, this signifies that the package contains stubs

> for the `foo==1.1.2`. In this example, the final element of the version

Reviewer: ```suggestion
for `foo==1.1.2`. In this example, the final element of the version
```

indicates that this is the first revision of the stubs for `foo`. If an update

> to the stubs was pushed (but the stubs were still aiming to provide

Reviewer: The English subjunctive ;)

```suggestion
to the stubs were pushed (but the stubs were still aiming to provide
```

annotations for `foo==1.1.2`), then the version of `types-foo` would
increment to `1.1.2.1`.

At typeshed, we try to keep breaking changes to a minimum. However, due to the
nature of stubs, any version bump can introduce changes that might make your
code fail to type check.

There are several strategies available for specifying the version of a stubs

> package you're using, each with its own tradeoffs

Reviewer 2: ```suggestion
package you're using, each with its own tradeoffs:
```

Reviewer: duplicate of https://github.com/python/typeshed/pull/10840#discussion_r1347533070 ;)

Reviewer 2: I spent too much time convincing myself that the "at least four parts" is right :) (Seems like pywin32 itself uses a two-part versioning scheme, but we add enough .0s to indeed make it four-part.)

> package you're using, each with its own tradeoffs

Reviewer: ```suggestion
package you're using, each with its own tradeoffs:
```

1. Use the same pin that you use for the package being stubbed. For example,
  if you use `requests>=2.30.0,<2.32`, you can use
  `types-requests>=2.30.0,<2.32`. This ensures that the stubs are compatible
  with the package you are using, but it carries a small risk of breaking

>   type checking due to changes in the stubs.

Reviewer: The indentation looks weird here, now that it's switched to a numbered list -- I think it all needs to be indented 3 spaces rather than two, now

  Another risk of this strategy is that stubs often lag behind
  the package being stubbed. You might want to force the package being stubbed
  to a certain minimum version, because it fixes a critical bug, but you
  could be unable to update the stubs, since an update has not been released.
2. Pin the stubs to a known good version and update manually. For example, if
  you use `types-requests==2.31.0.1`, you can be sure that upgrading
  dependencies will not break type checking. However, you will miss out on
  improvements in the stubs that could potentially improve type checking,

>   unless you manually update the stubs.

Reviewer: ```suggestion
  until you manually update the pin.
```

Also not sure about the word "manually" for this strategy, since many people use dependabot, renovate, or similar tools these days

> 2. Pin the stubs to a known good version and update manually. For example, if
>    you use `types-requests==2.31.0.1`, you can be sure that upgrading
>    dependencies will not break type checking. However, you will miss out on
>    improvements in the stubs that could potentially improve type checking,
>    until you update the pin.
>    It also has the risk that the stubs you are using are not compatible with
>    the package being stubbed.

Reviewer: ```suggestion
2. Pin the stubs to a known good version and update the pin from time to time
   (either manually, or using a tool such as dependabot or renovate).

   For example, if you use `types-requests==2.31.0.1`, you can have confidence
   that upgrading dependencies will not break type checking. However, you will
   miss out on improvements in the stubs that could potentially improve type
   checking until you update the pin. This strategy also has the risk that the
   stubs you are using might become incompatible with the package being stubbed.
```

> 3. Don't pin the stubs. This is the most flexible option, but it carries the
>   risk that the stubs become incompatible with the package being stubbed.

Reviewer: ```suggestion
3. Don't pin the stubs. This is the option that demands the least work from
   you when it comes to updating version pins, and has the advantage that you
   will automatically benefit from improved stubs whenever a new version of the
   stubs package is released. However, it carries the risk that the stubs
   become incompatible with the package being stubbed.

```

> 3. Don't pin the stubs. This is the most flexible option, but it carries the
>    risk that the stubs become incompatible with the package being stubbed.
>    For example, if a new major version of the package is released, there's a

Reviewer: ```suggestion
3. Don't pin the stubs. This is the option that demands the least work from
   you when it comes to updating version pins, and has the advantage that you
   will automatically benefit from improved stubs whenever a new version of the
   stubs package is released. However, it carries the risk that the stubs
   become incompatible with the package being stubbed.

   For example, if a new major version of the package is released, there's a
```

   chance the stubs might be updated to reflect the new version of the runtime
   package before you update the package being stubbed.

> You can also use the different strategies as needed. For example, you could
> default to strategy (1), but fall back to strategy (2) when

Reviewer: ```suggestion
You can also switch between the different strategies as needed. For example, you
could default to strategy (1), but fall back to strategy (2) when
```

Type checkers should be able to use these stub packages when installed. For more
details, see the documentation for your type checker.

### Package versioning for third-party stubs

Version numbers of third-party stub packages consist of at least four parts.
All parts of the stub version, except for the last part, correspond to the
version of the runtime package being stubbed. For example, if the `types-foo`
package has version `1.1.2.0`, this signifies that the package contains stubs
for `foo==1.1.2`. In this example, the final element of the version
indicates that this is the first revision of the stubs for `foo`. If an update
to the stubs were pushed (but the stubs were still aiming to provide
annotations for `foo==1.1.2`), then the version of `types-foo` would

> increment to `1.1.2.1`.

Reviewer: What happens if `foo` at runtime has version `1.1.1`, and typeshed has `version = 1.1.*` in `stubs/foo/METADATA.toml`? Will the version for `types-foo` be `1.1.0.whatever`, or `1.1.1.whatever`?

Reviewer: Hmm, looks like the version numbers can get out of sync in cases like that:
- `Deprecated` has version `2.1.14` at runtime: https://pypi.org/project/Deprecated/
- https://github.com/python/typeshed/blob/54a8fd74a785ff7916bade3699a2be4542f1b1c0/stubs/Deprecated/METADATA.toml#L1
- `types-Deprecated` is currently on version `2.1.9.3`: https://pypi.org/project/types-Deprecated/.

Reviewer: Maybe something like this?

```suggestion
version of the runtime package being stubbed. For example, if the `types-foo`
package has version `1.1.2.0`, this signifies that the package contains stubs
for `foo>=1.1.2`, but which have not been tested against `foo`>=1.2`.
In this example, the final element of the version indicates that this is the
first revision of the stubs for `foo>=1.1.2`. If an update to the stubs were
pushed (but the stubs were still aiming to provide annotations for
`foo>=1.1.2`), then the version of `types-foo` would increment to `1.1.2.1`.
```

Reviewer 3: There is some legacy there because of interaction with previous versioning. stub_uploader sees a) the specifier `"1.2.*"`, b) the fact that the last thing we uploaded was `1.2.9.3` (and at time of version swap last thing we uploaded was `1.2.9`) and just bumps.

The next time stubsabot bumps the version specifier, everything will become good. Specifier will be `1.3.*`, `Deprecated` will be `1.3.x` and we will be `1.3.0.y`.

Note the cause the complexity here is that we don't maintain multiple versions of stubs, so we do want to have each version of stub be workable for a specifier, but we can't put `*` into versions, so (assuming no legacy) we put `0` if possible

> `foo>=1.1.2`), then the version of `types-foo` would increment to `1.1.2.1`.

Reviewer: (Edited your suggestion to fix a minor typo)

> `foo>=1.1.2`), then the version of `types-foo` would increment to `1.1.2.1`.

Reviewer: I do still find the current situation with types-Deprecated that I described in https://github.com/python/typeshed/pull/10840#discussion_r1348644670  slightly hard to square with any explanation of our versioning scheme tbh. @hauntsaninja, any idea what's going on there?

> `foo>=1.1.2`), then the version of `types-foo` would increment to `1.1.2.1`.

Reviewer 3: Yeah, it's little confusing, commented on why it does what it does

At typeshed, we try to keep breaking changes to a minimum. However, due to the
nature of stubs, any version bump can introduce changes that might make your
code fail to type check.

There are several strategies available for specifying the version of a stubs
package you're using, each with its own tradeoffs:

1. Use the same pin that you use for the package being stubbed. For example,
   if you use `requests>=2.30.0,<2.32`, you can use
   `types-requests>=2.30.0,<2.32`. This ensures that the stubs are compatible
   with the package you are using, but it carries a small risk of breaking
   type checking due to changes in the stubs.

   Another risk of this strategy is that stubs often lag behind
   the package being stubbed. You might want to force the package being stubbed

>    to a certain minimum version, because it fixes a critical bug, but you
>    could be unable to update the stubs, since an update has not been released.

Reviewer 3: ```suggestion
   to a certain minimum version because it fixes a critical bug, but if
   correspondingly updated stubs have not been released, your type
   checking results may not be fully accurate.
```
nit: in practice, slightly stale stubs still usually work; I don't want to make it sound like you won't be able to get work done.
